### PR TITLE
Quick First Pass Implementation for Debugging Detail Calculations in the CLI

### PR DESCRIPTION
**Pull request by ctsims**
In the CLI you can now run

debug SELECTION_INDEX

on a select screen to get a breakdown of all of the variable and template calculations that run over that entity.

You can also run 

debug list

to expand the breakdown of how the nodeset itself was expanded.

Rough first pass. Have some good paths to refactoring this to generalize across the CLI

### CODE REVIEW

File: javarosa/src/main/java/org/javarosa/core/model/trace/AccumulatingReporter.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Vector;
+
+/**
+ * A simple accumulator which keeps track of expressions which are evaluated
+ * upon request.
+ *
+ * Created by ctsims on 10/19/2016.
+ */
+public class AccumulatingReporter implements EvaluationTraceReporter {
+    Vector<EvaluationTrace> traces = new Vector<>();
```
Comment (phillipm) [inline javarosa/src/main/java/org/javarosa/core/model/trace/AccumulatingReporter.java]: `private final`

File: util/src/org/commcare/util/cli/EntityListSubscreen.java [hunk 2]
```diff
@@ -43,11 +51,27 @@ public EntityListSubscreen(Detail shortDetail, Vector<TreeReference> references,
...
-        EvaluationContext context = new EvaluationContext(ec, entity);
+    private String createRow(TreeReference entity) {
+        return createRow(entity, false);
+    }
+
+    private String createRow(TreeReference entity, boolean collectDebug) {
+        EvaluationContext context = new EvaluationContext(rootContext, entity);
+        AccumulatingReporter reporter = new AccumulatingReporter();
 
+        if(collectDebug) {
```
Comment (phillipm) [inline util/src/org/commcare/util/cli/EntityListSubscreen.java]: Can you run the code format tool on your changes?

File: util/src/org/commcare/util/cli/EntityListSubscreen.java [hunk 3]
```diff
@@ -79,9 +103,25 @@ private String createRow(TreeReference entity, Detail shortDetail, EvaluationCon
...
+            if(reporter.getCollectedTraces().size() > 0) {
+                System.out.println("Template Traces:");
+            }
+            PrintCollectedTraces(reporter);
+            reporter.clearTraces();
+        }
         return row.toString();
     }
 
+    private static void PrintCollectedTraces(AccumulatingReporter reporter) {
```
Comment (phillipm) [inline util/src/org/commcare/util/cli/EntityListSubscreen.java]: `PrintCollectedTraces` -> `printCollectedTraces'

File: util/src/org/commcare/util/cli/EntityListSubscreen.java [hunk 4]
```diff
@@ -79,9 +103,25 @@ private String createRow(TreeReference entity, Detail shortDetail, EvaluationCon
...
             }
         }
+
+        if(collectDebug) {
+            if(reporter.getCollectedTraces().size() > 0) {
+                System.out.println("Template Traces:");
+            }
+            PrintCollectedTraces(reporter);
+            reporter.clearTraces();
+        }
```
Comment (phillipm) [inline util/src/org/commcare/util/cli/EntityListSubscreen.java]: This code is duplicated twice, can you lift into a helper method?

File: util/src/org/commcare/util/cli/EntityListSubscreen.java [hunk 5]
```diff
@@ -148,6 +188,15 @@ public boolean handleInputAndUpdateHost(String input, EntityScreen host) throws
             }
         }
 
+        if(input.startsWith("debug ")) {
```
Comment (phillipm) [inline util/src/org/commcare/util/cli/EntityListSubscreen.java]: I would expect this command to start with a `:` like `:back` and such

**Comment by ctsims** [comment]
@phillipm Changes processed. I couldn't update the debug command issue, because we catch ":" at a higher level, but I think it's something I want to refactor a lot in the future anyway.

**Comment by amstone326** [comment]
@phillipm I'm gonna merge this even though you haven't approved the requested changes because Clayton wanted to get this in for use by the enikshay team; hope that's ok!

**Pull request merged**